custom_metrics/wpt_bodies.js

```diff
@@ -0,0 +1,1366 @@
+    // Backup: heading order and length of h1s in almanac.js - not much
+    'headings': (() => { 
+      try {  
+      
+        function processHeading(d, target, n, level) {
+          let html  = n.innerHTML;
+          let text = seoText(n);
+
+          let words = text.match(/\S+/g)?.length;
+
+          if (!target.primary) {
+            if (text.length > 0) { // make this primary
+              let primaryTitle = "";
+              if (d == document) {
+                primaryTitle = renderedPrimaryTitle; 
+              }
+              else {
+                primaryTitle = rawPrimaryTitle; 
+              }
+
+              let snippet = text;
+
+              if (snippet.length > 100) {
+                snippet = snippet.substring(0, 100)+"...";
+              }
+
+              target.primary = {
+                words: words,
+                characters: text.length,
+                matches_title: text.toLowerCase() == primaryTitle?.toLowerCase(),
+                text: snippet,
+                level: level
+              }
+            } else if (html.length > 0) {
+              // looks like a hidden heading
+              target.first_non_empty_heading_hidden = true;
+            }
+          }
+          return {characters: text.length, words: words ?? 0};
+        }
+        
+        function processHeadings(d) {
+          let target = {first_non_empty_heading_hidden: false};
+          for(let l=1; l < 9; l++) {
+            let nodes = Array.from(d.querySelectorAll('h'+l));
+
+            let characters = 0;
+            let words = 0;
+            // if don't have a primary heading yet, search for one.
+            
+            var hs = nodes.map(n => {
+              let h = processHeading(d, target, n, l);
+              characters += h.characters;
+              words += h.words;
+              return h;
+            });
+      
+            target["h"+l] = {
+              total: nodes.length,
+              non_empty_total: nodes.filter(e => seoText(e).length > 0).length,
+              characters: characters,
+              words: words
+            };
+          }
+          return target;
+        }
+
+        let result = {};
+
+        result.rendered = processHeadings(document);
+
+        let rawHtmlDocument = getRawHtmlDocument();
+
+        if (rawHtmlDocument) {
+          result.raw = processHeadings(rawHtmlDocument);
+        }
+
+
+    
+        return result; 
+      }
+      catch(e) {
+        return logError("headings", e);
+      }
+    })(),
+
+    // Structured Data use
+    // Used by: SEO
+    // Backup: old data in almanac.js - not very good
+    'structured_data': (() => {  
+      try { 
+        var link = document.createElement('a');
+
+        function nestedJsonldLookup(target, jsonldIds, items, depth, context) {
+          if (items instanceof Array) {
+            // loop array and process any objects in it 
+            for (var i = 0, len = items.length; i < len; i++) {
```

Comment:
`items.forEach`?

Reply:
Performance is mostly negligible, I suggested for consistency and simplicity, but up to you to resolve as you see fit.

Reply:
I bit the bullet and refactored forEach and array related stuff.